CONTRIBUTING.md

```diff
@@ -137,6 +137,56 @@ maximize the chances of your PR being merged.
   [envoy-filter-example](https://github.com/envoyproxy/envoy-filter-example) (for example making a new
   branch so that CI can pass) it is your responsibility to follow through with merging those
   changes back to master once the CI dance is done.
+* If your PR is a high risk change, the reviewer may ask that you runtime guard
+  it. See the section on runtime guarding below.
+
+
+# Runtime guarding
+
+Some high risk changes in Envoy are deemed worthy of runtime guarding. Instead of just replacing
+old code with new code, both code paths are supported for between one Envoy release (if it is
+guarded due to performance concerns) and a full deprecation cycle (if it is a high risk behavioral
+change).
+
+The canonican way to runtime guard a feature is
+```
+if (Runtime::LoaderSingleton::getExisting() &&
+    Runtime::LoaderSingleton::getExisting()->snapshot()->runtimeFeatureEnabled("
+       envoy.reloadable_features.my_feature_name")) {
+  [new code path]
+} else {
+  [old_code_path]
+}
+```
+Runtime guarded features named with the "envoy.reloadable_features." prefix must be safe to flip
+true or false on running Envoy instances. In some situations, for example the buffer rewrite in
+[#5441](https://github.com/envoyproxy/envoy/pull/5441), it may make more sense to
+latch the value in a member variable on class creation, for example:
+
+```
+bool use_new_code_path_ = Runtime::LoaderSingleton::getExisting() &&
+  Runtime::LoaderSingleton::getExisting()->snapshot()->runtimeFeatureEnabled(
+      "envoy.reloadable_features.my_feature_name"));
```

Comment:
Is it worth trying to make this a bit less verbose at the sites where it is used? There's a lot of boilerplate to get to `runtimeFeatureEnabled`.

Reply:
Yeah we could likely have a wrapper that internally loads the singleton and does the right thing if it doesn't exist.

Reply:
Definitely the right way to go.

Less clear, should we use this in tests?  I was a bit bummed at the heavyweight work in TestEnvironment to be able to flag override - setting up all those fake components.  With the helper if we want to have a friend test-only class in ConstSingleton we can skip all the static state and just muck with the const singleton directly.  I think that might also work for integration tests, which would be pretty awesome.

That said, it's making our const-singleton non-const and I know how much the larger Envoy community _loves_ hacks like that ;-)

Reply:
What about an approach like https://github.com/envoyproxy/envoy/pull/6139 with some test class mixin that changes the singleton behavior? That seems like a really cleanly structured way to have a "mostly const" singleton that alters behavior only within a particular test.